File: nim/day12.nim
import re, strutils, json

const instructions = readFile("./inputs/12.txt").string
let pattern = re"-?\d+\b"
var total: int

for m in re.findAll(instructions, pattern):
  total += m.parseInt

echo total


let a = parseJson(instructions)

func calc(a: JsonNode): int =
  case a.kind
  of JInt:
    result += a.getInt
  of JArray:
    for e in a.items:
      result += calc(e)
  of JObject:
    var partial = 0
    for k, v in a.pairs:
      if v.kind == JString and "red" in v.str:
        return
      partial += calc(v)
    result += partial
  else: discard

echo calc(a)

File: python/day12.py
import json
import re


numbers = re.compile("-?\d+")

def find_sum(jason):
    if type(jason) is int:
        return jason
    if type(jason) is dict:
        if "red" in jason.values(): return 0
        else: return sum(map(find_sum, jason.values()))
    if type(jason) is list:
        return sum(map(find_sum, jason))
    else: return 0


data = open("inputs/12.txt").read()
j = json.loads(data)

print(sum(map(int, re.findall(numbers, data))))
print(find_sum(j))
